cloud/aws/actuators/machine/actuator.go

```diff
@@ -183,6 +243,408 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+func (a *Actuator) ensureTags(machine *clusterv1.Machine, instanceID *string, additionalTags map[string]string) (bool, error) {
+	annotation, err := a.machineAnnotationJSON(machine, TagsLastAppliedAnnotation)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the instance tags were changed. If they were, update them.
+	// It would be possible here to only send new/updated tags, but for the
+	// moment we send everything, even if only a single tag was created or
+	// upated.
+	changed, created, deleted, newAnnotation := a.tagsChanged(annotation, additionalTags)
+	if changed {
+		err = a.ec2.UpdateResourceTags(instanceID, created, deleted)
+		if err != nil {
+			return false, err
+		}
+
+		// We also need to update the annotation if anything changed.
+		err = a.updateMachineAnnotationJSON(machine, TagsLastAppliedAnnotation, newAnnotation)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return changed, nil
+}
+
+// Arguments:
+//  - annotation: The security group annotation.
+//  - src: The Kubernetes state.
+//  - dst: The AWS state.
+// We define them this way since we are going to change the dst (AWS) based on
+// the src (Kubernetes).
+//
+// Returns:
+//  - bool: Indicates if the state changed.
+//  - []*string{}: Array of securityGroupIDs, these should be passed to AWS
+//    if the state changed.
+//  - map[string]interface{}: The new annotation to set on the machine.
+//
+// First, we are comparing the annotation to the src to see if any security
+// groups were removed from the src since the last time Update was called.
+//
+// Based on this we create a new map[string]bool, representing groupID=>present.
+// groupID=>true indicates a new or already existing groupID.
+// groupID=>false indicates a groupID that was removed since last time we were
+// called.
+//
+// Once we have the above map representing the changes since we last ran, we
+// start to prepare a map, representing what we'd like state in AWS to look
+// like.
+// We first add the groupIDs from our above state that need to be kept.
+// We then compare the state in AWS to our above state.
+//  - If we find a groupID in AWS that isn't in our state, that is a security
+// 	  group that we don't manage. We add it to our new map with no further
+//    checking.
+//  - If we find a groupID in AWS that's in our state and flagged to be kept,
+//    we add it to the map.
+//  - If we find a groupID in AWS that's in our state and flagged for deletion,
+//    we do not add it to the map and we allow it to drop.
+//
+// Next we compare the map we have made after comparing our desired state and
+// the AWS state against AWS again to decide if any changes actually happened.
+//
+// Finally, we return three items listed at the top of this comment.
+// The caller then acts upon this information.
+func (a *Actuator) securityGroupsChanged(annotation map[string]interface{}, src []v1alpha1.AWSResourceReference, dst map[string]string) (bool, []*string, map[string]interface{}) {
```

Comment:
I think we will want a set package. It's ok to put this off for now, but the bulk of this function should only be a few lines doing set difference and intersection and such.